File: src/events.ts
import { isFunction } from './utilities'

export type EventEmitterCallback<T = any> = {
    (value?: T): void
    originFunction?: EventEmitterCallback<T>
}

export class EventEmitter<Scope = any> {
    private readonly _events: Map<string, EventEmitterCallback[]>
    private readonly _scope?: Scope

    constructor(scope?: Scope) {
        this._events = new Map<string, EventEmitterCallback[]>()
        if (scope) this._scope = scope
    }

    public on<T>(eventName: string, callback: EventEmitterCallback<T>): void {
        if (!this._events.has(eventName)) this._events.set(eventName, [])
        this._events.get(eventName).push(callback)
    }

    public emit<T>(eventName: string, value?: T): void {
        if (!this._events.has(eventName)) return
        this._events.get(eventName).forEach((callback) => {
            if (isFunction(callback)) {
                if (this._scope) callback.call(this._scope, value)
                else callback(value)
            }
        })
    }

    public off<T>(eventName: string, callback?: EventEmitterCallback<T>): void {
        if (callback) {
            this._events.set(
                eventName,
                this._events.get(eventName).filter((cb) => {
                    if (cb === callback || cb.originFunction === callback)
                        return false
                })
            )
        } else {
            this._events.delete(eventName)
        }
    }

    public once<T>(
        eventName: string,
        callback?: EventEmitterCallback<T>
    ): void {
        const self = this
        const onceCallback = function () {
            self.off(eventName, onceCallback)
            callback.apply(self, arguments)
        }
        onceCallback.originFunction = callback
        this.on(eventName, onceCallback)
    }
}

File: lib/events.go
package lib

import "reflect"

type Callback func(interface{})

type EventEmitterCallback struct {
	Callback
	OriginCallback *EventEmitterCallback
}

type EventEmitter struct {
	events map[string][]*EventEmitterCallback
}

func NewEventEmitterCallback(callback Callback) *EventEmitterCallback {
	return &EventEmitterCallback{Callback: callback}
}

func NewEventEmitter() *EventEmitter {
	e := new(EventEmitter)
	e.events = make(map[string][]*EventEmitterCallback, 100)
	return e
}

func (e *EventEmitter) Events() map[string][]*EventEmitterCallback {
	return e.events
}

func (e *EventEmitter) On(eventName string, callback Callback) {
	if _, ok := e.events[eventName]; !ok {
		e.events[eventName] = []*EventEmitterCallback{}
	}
	ec := NewEventEmitterCallback(callback)
	e.events[eventName] = append(e.events[eventName], ec)
}

func (e *EventEmitter) Emit(eventName string, value interface{}) {
	eCallbacks, ok := e.events[eventName]
	if !ok {
		return
	}
	for _, cb := range eCallbacks {
		cb.Callback(value)
	}
}

func Filter(stu []*EventEmitterCallback, f func(s interface{}) bool) []*EventEmitterCallback {
	var r []*EventEmitterCallback
	for _, s := range stu {
		if f(s) == true {
			r = append(r, s)
		}
	}
	return r
}

func (e *EventEmitter) Off(eventName string, callbacks ...Callback) {
	if len(callbacks) != 0 {
		e.events[eventName] = Filter(e.events[eventName], func(v interface{}) bool {
			eCallback := v.(*EventEmitterCallback)
			var eOriginCallbackPointer uintptr
			eCallbackPointer := reflect.ValueOf(eCallback.Callback).Pointer()
			if eCallback.OriginCallback != nil {
				eOriginCallbackPointer = reflect.ValueOf(eCallback.OriginCallback.Callback).Pointer()
			}
			for _, callback := range callbacks {
				callbackPointer := reflect.ValueOf(callback).Pointer()
				if eCallbackPointer == callbackPointer || eOriginCallbackPointer == callbackPointer {
					return false
				}
			}
			return true
		})
	} else {
		delete(e.events, eventName)
	}
}

func (e *EventEmitter) Once(eventName string, callback Callback) {
	var onceCallback *EventEmitterCallback
	onceCallback = NewEventEmitterCallback(func(value interface{}) {
		e.Off(eventName, onceCallback.Callback)
		callback(value)
	})
	onceCallback.OriginCallback = NewEventEmitterCallback(callback)
	if _, ok := e.events[eventName]; !ok {
		e.events[eventName] = []*EventEmitterCallback{}
	}
	e.events[eventName] = append(e.events[eventName], onceCallback)
}
